crates/parity-util-mem/src/allocators.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Copyright 2020 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[cfg(feature = "std")]
+use crate::malloc_size::MallocUnconditionalSizeOf;
+use crate::malloc_size::{MallocSizeOf, MallocSizeOfOps, VoidPtrToSizeFn};
+#[cfg(not(feature = "std"))]
+use core::ffi::c_void;
+#[cfg(feature = "std")]
+use std::os::raw::c_void;
+
+mod usable_size {
+
+    use super::*;
+
+    cfg_if::cfg_if! {
+
+        if #[cfg(any(
+            target_arch = "wasm32",
+            feature = "estimate-heapsize",
+        ))] {
+
+            // do not try system allocator
+
+            /// Warning this is for compatibility only.
+            /// This function does panic: `estimate-heapsize` feature needs to be activated
+            /// to avoid this function call.
+            pub unsafe extern "C" fn malloc_usable_size(_ptr: *const c_void) -> usize {
+                unreachable!("estimate heapsize only")
+            }
+
+        } else if #[cfg(target_os = "windows")] {
+
+            use winapi::um::heapapi::{GetProcessHeap, HeapSize, HeapValidate};
+            use winapi::ctypes::c_void as winapi_c_void;
+
+            /// Get the size of a heap block.
+            /// Call windows allocator through `winapi` crate
+            pub unsafe extern "C" fn malloc_usable_size(mut ptr: *const c_void) -> usize {
+
+                let heap = GetProcessHeap();
+
+                if HeapValidate(heap, 0, ptr as *const winapi_c_void) == 0 {
+                    ptr = *(ptr as *const *const c_void).offset(-1);
+                }
+
+                HeapSize(heap, 0, ptr as *const winapi_c_void) as usize
+            }
+
+        } else if #[cfg(any(
+            target_os = "linux",
+            target_os = "android",
```

Comment:
Surely we don't need this? :)

Reply:
Oh yeah, I tried to remove all that crap that I could find but thanks for pointing this out.

Reply:
Done although it doesn't seem to be related to the windows rocksdb build failure.

(I made `estimate-heapsize` enabled by default so I think that should never actually be selected in any of our builds)